### PR TITLE
Add support for indexers in assign param

### LINKED ISSUE #42082: Add back indexer expression support
Indexer expression support was removed in https://github.com/Azure/azure-sdk-for-net/commit/51102d0b376f77838a11ed8a34594ae7642bcc5b#diff-7b26e8e1b26b128c5fea766c170364956f8872545807cc43829fdc0286f13adaL85-L90

We should add it back, and then update KeyVaultAddAccessPolicy to use AssignParameter rather than using Parameter/ParameterOverrides/AddScope.

We will then need to update this assertion to expect 1 - https://github.com/Azure/azure-sdk-for-net/blob/feature/cdk/sdk/provisioning/Azure.Provisioning/tests/ProvisioningTests.cs#L175

### PR DESCRIPTION
**Pull request by m-nash**
Fixes https://github.com/Azure/azure-sdk-for-net/issues/42082

# Contributing to the Azure SDK

Please see our [CONTRIBUTING.md](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md) if you are not familiar with contributing to this repository or have questions.

For specific information about pull request etiquette and best practices, see [this section](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md#pull-request-etiquette-and-best-practices).

### CODE REVIEW

File: sdk/provisioning/Azure.Provisioning/src/sqlmanagement/SqlDatabase.cs [hunk 1]
```diff
@@ -19,11 +19,12 @@ public class SqlDatabase : Resource<SqlDatabaseData>
         /// Initializes a new instance of the <see cref="SqlDatabase"/>.
         /// </summary>
         /// <param name="scope">The scope.</param>
+        /// <param name="parent">The parent.</param>
         /// <param name="name">The name.</param>
         /// <param name="version">The version.</param>
         /// <param name="location">The location.</param>
-        public SqlDatabase(IConstruct scope, string name = "db", string version = "2022-08-01-preview", AzureLocation? location = default)
-            : base(scope, null, name, ResourceTypeName, version, (name) => ArmSqlModelFactory.SqlDatabaseData(
+        public SqlDatabase(IConstruct scope, SqlServer? parent = null, string name = "db", string version = "2022-08-01-preview", AzureLocation? location = default)
```
Comment (JoshLove-msft) [inline sdk/provisioning/Azure.Provisioning/src/sqlmanagement/SqlDatabase.cs:26]: intentional?

File: sdk/provisioning/Azure.Provisioning/tests/ConstructTests.cs [hunk 2]
```diff
@@ -110,7 +110,7 @@ public void GetParametersChildConstructs(bool recursive)
             var rg2 = new ResourceGroup(childScope, "rg2");
             rg2.AssignParameter(r => r.Location, new Parameter("location"));
 
-            var expected = recursive ? 2 : 1;
+            var expected = recursive ? 3 : 2;
```
Comment (JoshLove-msft) [inline sdk/provisioning/Azure.Provisioning/tests/ConstructTests.cs:113]: Don't we need to update the assertion that was expecting 0 as well?

File: sdk/provisioning/Azure.Provisioning/src/sqlmanagement/SqlDatabase.cs (hunk 1 above)
Comment (m-nash) [reply sdk/provisioning/Azure.Provisioning/src/sqlmanagement/SqlDatabase.cs:26]: yeah, adding in what mitch mentioned as an addon to this pr :-)

File: sdk/provisioning/Azure.Provisioning/tests/ConstructTests.cs (hunk 2 above)
Comment (m-nash) [reply sdk/provisioning/Azure.Provisioning/tests/ConstructTests.cs:113]: The param is added to the keyvault's scope which is the infra which is why it went up by one but the l2 construct stays 0

**Pull request merged**